### PR TITLE
Add exception type coherence check on Haskell side

### PR DESCRIPTION
**Pull request by sofiafaro-da**
changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs [hunk 1]
```diff
@@ -306,6 +313,12 @@ instance Pretty Error where
       "expected a type containing neither type variables nor quantifiers, but found: " <> pretty foundType
     EExpectedExceptionType foundType ->
       "expected an exception type, but found: " <> pretty foundType
+    EExpectedExceptionTypeHasNoParams modName exnName ->
+      "exception type should not have type parameters: " <> pretty modName <> "." <> pretty exnName
```
Comment (hurryabit) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs:317]: "should" sounds like a friendly recommendation. Maybe "must" is more appropriate here?

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs [hunk 2]
```diff
@@ -829,14 +831,25 @@ checkTemplateKey param tcon TemplateKey{..} = do
...
 checkModule :: MonadGamma m => Module -> m ()
-checkModule m@(Module _modName _path _flags synonyms dataTypes values templates _exceptions) = do
+checkModule m@(Module _modName _path _flags synonyms dataTypes values templates exceptions) = do
   let with ctx f x = withContext (ctx x) (f x)
   traverse_ (with (ContextDefTypeSyn m) checkDefTypeSyn) synonyms
   traverse_ (with (ContextDefDataType m) checkDefDataType) dataTypes
   traverse_ (with (\t -> ContextTemplate m t TPWhole) $ checkTemplate m) templates
   traverse_ (with (ContextDefValue m) checkDefValue) values
-  -- TODO #8020: check exception types are well formed and don't overlap with templates
+  traverse_ (with (ContextDefException m) (checkDefException m)) exceptions
```
Comment (hurryabit) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs:855]: It's probably better to move this before checking any expressions such that we've run through this before we call `checkExceptionType`. It's not strictly necessary but I think it makes for better error messages.

**Review by hurryabit — APPROVED** [review]
Nice. Thank you.

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs (hunk 2 above)
Comment (sofiafaro-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs:855]: Oh, good catch!

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs (hunk 1 above)
Comment (sofiafaro-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs]: 👍  changed!

**Pull request merged**